lib/Dialect/TTIR/IR/TTIROps.cpp

```diff
@@ -1654,23 +1655,78 @@ static void createReduceOp(::mlir::OpBuilder &opBuilder, ::mlir::Block *block,
   opBuilder.create<mlir::tt::ttir::YieldOp>(loc, kernelOp->getResults());
 }
 
-// Sum op kernel builder
-void mlir::tt::ttir::SumOp::buildGenericRegion(::mlir::OpBuilder &opBuilder,
+// Common verifier for all Reduce ops.
+static mlir::LogicalResult
+verifyReduceOp(mlir::Operation *reduceOp,
```

Comment:
Because that way the code inside the template will be recompiled for each of the Reduce ops, adding unnecessary bloat to the binary. I am all for using templates where appropriate, but the pattern where you extract template code into separate function is very useful to avoid the recompilation and also to be able to implement that function in the .cpp file and hide its internals from the header (templates declared in headers can't be defined in cpp files).

> there's no need for unnecessary and potentially dangerous casts

You are referencing the cast of the first operand type to `RankedTensorType`? I can pass the input as the parameter of the function, but I wanted to save the number of parameters required, since we are already passing the op. It doesn't really seem dangerous, reduce ops will always have at least one operand, but I can do it if you insist :)

Reply:
Yeah, template instantiation crossed my mind, though in this particular case I didn't see it as a problem, as the number of instantiations is bounded and small. Still, I agree it's something to consider in general case.

When you add verification for shapes you will also need to communicate `keepDim`, that's why I find template solution a sweet spot between correctness, performance and convenience in this case. I'm also okay with `(RankedTensorType inputType, RankedTensorType outputType, const std::optional<mlir::ArrayAttr> &reduceDims, const bool keepDim)` as well. With `Operation *` you save on one parameter, but you still need to pass `reduceDims` and `keepDim` explicitly. Plus as casts happen dynamically they are not free either.

When I say dangerous I mean generally, I'm trying to treat every cast as dangerous, because it's something that happens during runtime. If I know the type of value during the writing of the code I like to communicate that knowledge to the compiler. Sometimes casts are necessary and otherwise would require a lot of boilerplate to circumvent them, so I'm not against using them when appropriate, I'm just seeing it in a lot of places in our code where they are totally unnecessary.

Reply:
Another aspect to consider when it comes to templates is unit testing. If we wanted to unit test that template, we would need to move it to the .h file in order to access it, so its definition would have to be in the .h file, and all the internals inside would be exposed with the header including additional includes that are required for those internals.  Passing a few additional parameters to the function is not really a big deal.

I've added additional parameter for `inputType` to avoid casting.